snapshots/windows/common.go

```diff
@@ -0,0 +1,419 @@
+//go:build windows
+// +build windows
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package windows
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/Microsoft/hcsshim"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/mount"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/containerd/containerd/snapshots/storage"
+	"github.com/containerd/continuity/fs"
+)
+
+// windowsBaseSnapshotter is a type that implements common functionality required by both windows & cimfs
+// snapshotters (sort of a base type that windows & cimfs snapshotter types derive from - however, windowsBaseSnapshotter does NOT impelement the full Snapshotter interface).  Some functions
+// (like Stat, Update) that are identical for both snapshotters are directly implemented in this base
+// snapshotter and such functions handle database transaction creation etc. However, the functions that are
+// not common don't create a transaction to allow the caller the flexibility of deciding whether to commit or
+// abort the transaction.
+type windowsBaseSnapshotter struct {
+	root string
+	ms   *storage.MetaStore
+	info hcsshim.DriverInfo
+}
+
+func newBaseSnapshotter(root string) (*windowsBaseSnapshotter, error) {
+	if err := os.MkdirAll(root, 0700); err != nil {
+		return nil, err
+	}
+
+	ms, err := storage.NewMetaStore(filepath.Join(root, "metadata.db"))
+	if err != nil {
+		return nil, err
+	}
+
+	if err := os.Mkdir(filepath.Join(root, "snapshots"), 0700); err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+
+	return &windowsBaseSnapshotter{
+		root: root,
+		ms:   ms,
+		info: hcsshim.DriverInfo{HomeDir: filepath.Join(root, "snapshots")},
+	}, nil
+}
+
+func (w *windowsBaseSnapshotter) getSnapshotDir(id string) string {
+	return filepath.Join(w.root, "snapshots", id)
+}
+
+func (w *windowsBaseSnapshotter) parentIDsToParentPaths(parentIDs []string) []string {
+	var parentLayerPaths []string
```

Comment:
Since you know the size, you can do:
```suggestion
	parentLayerPaths := make([]string, 0, len(parentIDs))
```